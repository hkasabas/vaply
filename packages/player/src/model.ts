export type NodePlayerConfig = NodePackageConfig & {
  flow: NodeFlowConfig;
  nodes: Array<
    ContentNodeConfig & {
      annotations: Array<AnnotationConfig>;
      // triggers connecting content movement and annotations
      triggers: Array<AnnotationTriggerConfig>;
    }
  >;
};

export type NodePackageConfig = {
  id: string;
  title: string;
};

export type NodeFlowType = "list";
export type NodeFlowConfig = {
  type: NodeFlowType;
  // [<NodeFlowType>]: {},
  list: {
    nodes: string[];
  };
};

export type ContentNodeConfig = {
  code: string; // custom generated ID for internal referencing (eg. positioning)

  video: VideoContentConfig;
};

export type VideoContentType = "file";
export interface VideoContentConfig {
  type: VideoContentType;

  // [<VideoContentType>]: {},
  file?: FileVideoContentConfig;
}

export interface FileVideoContentConfig {
  src: string;
  poster?: string;
}

export type AnnotationType = "card" | "externalcontent" | "chapter";
export type AnnotationPositionType = "static" | "none";
export type AnnotationConfig = {
  code: string;
  type: AnnotationType;
  blocking?: boolean;
  dismissible?: boolean;

  position: {
    type: "static";

    static: Partial<{
      top: string; // CSS dimensions
      right: string; // CSS dimensions
      bottom: string; // CSS dimensions
      left: string; // CSS dimensions
    }>;
  };

  dimensions?: Partial<{
    width: string; // CSS dimensions
    height: string; // CSS dimensions
  }>;
} & (
  | { type: "card"; card: CardAnnotationConfig }
  | { type: "externalcontent"; externalcontent: ExternalContentAnnotationConfig }
  | { type: "chapter"; chapter: ChapterAnnotationConfig }
);

export type CardAnnotationConfig = {
  title?: string;
};
export type ExternalContentAnnotationConfig = {
  selector: string;
};
export type ChapterAnnotationConfig = {
  title: string;
};

export type AnnotationTriggerType = "timeupdate" | "route";
export type AnnotationTriggerConfig =
  | {
      targetCode: string; // annotation code
      type: "timeupdate";
      timeupdate: {
        start: string | number;
        end?: string | number;
      };
    }
  | {
      targetCode: string; // annotation code
      type: "route";
      route: {
        start: string | number;
      };
    };

// ---------- events

export type RouteEvent = string;

export type ObjectValue<T> = { value: T | undefined };
