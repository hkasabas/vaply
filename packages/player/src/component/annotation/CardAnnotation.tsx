import { FunctionComponent, h } from "preact";

import { AnnotationConfig, DataEvent, RouteEvent, TypedUnion } from "@player/model";

/** Card annotation component props */
export type CardAnnotationProps = {
  config: TypedUnion<AnnotationConfig, "card">;

  onData?: (data: DataEvent) => void;
  onRoute?: (route: RouteEvent) => void;
  onClose?: () => void;
};

/** Card annotation component */
const CardAnnotation: FunctionComponent<CardAnnotationProps> = (props) => {
  const cardConfig = props.config.card;

  return <div className="vaply-cardAnnotation__container">{cardConfig.title}</div>;
};

export { CardAnnotation };
