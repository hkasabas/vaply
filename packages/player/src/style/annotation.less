
// ---------- Common

.vaply-annotation__staticPosition {
  position: absolute;
  z-index: 10;

  .vaply-annotation__closeButton {
    position: absolute;
    top:0px;
    right: 0px;
    font-weight: bold;
    cursor: pointer;
    font-size: 1.1rem;
    line-height: 1.1rem;
    width: 1.1rem;
    text-align: center;
  }
}


// ---------- Card annotation

.vaply-cardAnnotation__container {
  position: relative;
  background-color: white;
  opacity: .5;
  padding: 0.8rem 1rem;
  width: calc(100% - 2*1rem);
  height: calc(100% - 2*0.8rem);
}

