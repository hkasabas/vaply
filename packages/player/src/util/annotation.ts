import { AnnotationConfig, AnnotationTriggerConfig, AnnotationTriggerType } from "@player/model";
import { assertUnreachable } from "@player/util/utils";

/** Typeguard fn that TS compiler understands when used in filters. */
function isNotEmpty<T>(obj: T): obj is NonNullable<T> {
  return obj != null;
}

export function findApplicableAnnotations(
  position: string,
  type: AnnotationTriggerType,
  availableAnnotations: AnnotationConfig[],
  triggers: AnnotationTriggerConfig[]
): AnnotationConfig[] {
  return (
    triggers
      // trigger by type
      .filter((trigger) => {
        return trigger.type === type;
      })
      // trigger by time
      .filter((trigger) => {
        return isActiveAnnotation(position, trigger);
      })
      // map trigger to annotation
      .map((trigger) => {
        return availableAnnotations.find((ann) => ann.code === trigger.targetCode);
      })
      // remove empty - triggers with targets not found in
      .filter(isNotEmpty)
  );
}

/** Remove previously closed annotations from given available annotation list. */
export function findActiveAnnotations(availableAnnotations: AnnotationConfig[], closedAnnotations: AnnotationConfig[]): AnnotationConfig[] {
  return (
    availableAnnotations
      // remove already closed
      .filter((cAnnot) => {
        return closedAnnotations.find((annot) => annot.code == cAnnot.code) == null;
      })
  );
}

/** Helper function which checks if there are any blocking annotations in given annotation list. */
export function hasBlockingAnnotations(availableAnnotations: AnnotationConfig[]): boolean {
  return (
    availableAnnotations
      // is there at least one that is blocking
      .some((annot) => annot.blocking)
  );
}

/** Return "start" specific for trigger type. */
export function findTriggerStart(trigger: AnnotationTriggerConfig): string | number | undefined {
  const type = trigger.type;
  switch (type) {
    case "route":
      return trigger.route.start;
    case "timeupdate":
      return trigger.timeupdate.start;
    default:
      assertUnreachable(type);
  }
}

// ----------

function isActiveAnnotation(position: string, trigger: AnnotationTriggerConfig): boolean {
  if (trigger.type === "timeupdate") {
    // in case of timeupdate annotations, position is number
    const timeNumber = Number.parseInt(position, 10);
    if (!Number.isNaN(timeNumber)) {
      // TODO: remove "as any"
      return (
        trigger.timeupdate != null &&
        (trigger.timeupdate.start as any) <= timeNumber &&
        (trigger.timeupdate.end == null || (trigger.timeupdate.end as any) > timeNumber)
      );
    }
    // NaN? no need to worry. This could be one of special routes that we just don't handle here yet
  }

  return false;
}
