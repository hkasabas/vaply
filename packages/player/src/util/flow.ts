import { NodeFlowConfig } from "@player/model";

export function getFirstNode(config: NodeFlowConfig): string | undefined {
  // default to "list" flow
  return getListFirstNode(config.list.nodes);
}

export function getPreviousNode(
  nodeId: string,
  config: NodeFlowConfig
): string | undefined {
  // default to "list" flow
  return getListPreviousNode(nodeId, config.list.nodes);
}

export function getNextNode(
  nodeId: string,
  config: NodeFlowConfig
): string | undefined {
  // default to "list" flow
  return getListNextNode(nodeId, config.list.nodes);
}

/**
 * List node flow
 *
 * List flow keeps nodes in a list and executes them sequentially.
 */

function getListFirstNode(nodes: string[]): string | undefined {
  return nodes.slice(0, 1).shift();
}

function getListPreviousNode(
  nodeId: string,
  nodes: string[]
): string | undefined {
  let previousId: string | undefined;
  let foundId: string | undefined;

  for (let i = 0; i < nodes.length; i++) {
    const id = nodes[i];
    if (id === nodeId) {
      foundId = previousId;
      break;
    }
    previousId = id;
  }

  return foundId;
}

function getListNextNode(nodeId: string, nodes: string[]): string | undefined {
  let previousId: string | undefined;
  let foundId: string | undefined;

  // we're looking for the next node so we'll iterate the arr in reverse to keep the same logic
  for (let i = nodes.length - 1; i >= 0; i--) {
    const id = nodes[i];
    if (id === nodeId) {
      foundId = previousId;
      break;
    }
    previousId = id;
  }

  return foundId;
}
