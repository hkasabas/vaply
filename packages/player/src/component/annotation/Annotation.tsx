import { h, FunctionComponent } from "preact";
import { useCallback, useMemo } from "preact/hooks";

import { CardAnnotation } from "@player/component/annotation/CardAnnotation";
import { AnnotationConfig, DataEvent, RouteEvent } from "@player/model";

/** Annotation component props */
export type AnnotationProps = {
  config: AnnotationConfig;

  onData?: (data: DataEvent) => void;
  onRoute?: (route: RouteEvent) => void;
  onClose?: () => void;
};

/**
 * Annotation component
 *
 * Display annotation config using appropriate annotation component
 */
const Annotation: FunctionComponent<AnnotationProps> = (props) => {
  // annotation positioning
  const positionProps = useMemo(() => {
    return {
      top: props.config.position.static?.top,
      right: props.config.position.static?.right,
      bottom: props.config.position.static?.bottom,
      left: props.config.position.static?.left,
      width: props.config.dimensions?.width,
      height: props.config.dimensions?.height,
    };
  }, [props.config.position.static]);

  const handleClose = useCallback(() => {
    props.onClose?.();
  }, [props.onClose]);

  return (
    <div className="vaply-annotation__staticPosition" style={positionProps}>
      {/* ----- card ----- */}
      {props.config.type === "card" && <CardAnnotation config={props.config} onData={props.onData} onRoute={props.onRoute} onClose={props.onClose} />}

      {/* close button */}
      {props.config.dismissible && (
        <div className="vaply-annotation__closeButton" onClick={handleClose}>
          &times;
        </div>
      )}
    </div>
  );
};

export { Annotation };
