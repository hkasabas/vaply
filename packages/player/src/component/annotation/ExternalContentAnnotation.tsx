import { FunctionComponent, h } from "preact";
import { useEffect, useRef } from "preact/hooks";

import { AnnotationConfig, RouteEvent } from "@player/model";
import { FilteredByType } from "@player/util/unionFilter";

/** External content annotation component props */
export interface ExternalContentAnnotationProps {
  config: FilteredByType<AnnotationConfig, "externalcontent">;

  onRoute?: (route: RouteEvent) => void;
  onClose?: () => void;
}

/** External content annotation component */
const ExternalContentAnnotation: FunctionComponent<ExternalContentAnnotationProps> = (props) => {
  const config = props.config.externalcontent;

  const containerRef = useRef<HTMLDivElement>(null);
  const externalContentRef = useRef<Element | null>(null);
  const externalContentParentRef = useRef<Element | null>(null);

  useEffect(() => {
    const el = document.querySelector(config.selector);
    const parentEl = el?.parentElement;

    if (el != null && parentEl != null) {
      // --- store content el and it's original parent
      externalContentRef.current = el;
      externalContentParentRef.current = el.parentElement;

      // detach element
      el?.parentElement?.removeChild(el);

      containerRef.current?.appendChild(el);
    } else {
      console.warn(`Empty external el or it's parent`, el, parentEl);
    }

    return () => {
      if (externalContentRef.current && externalContentParentRef.current) {
        // reattach element to it's original parent
        externalContentParentRef.current.appendChild(externalContentRef.current);
      }

      externalContentRef.current = null;
      externalContentParentRef.current = null;
    };
  }, [config.selector]);

  return <div className="vaply-externalContentAnnotation__container" ref={containerRef} />;
};

export { ExternalContentAnnotation };
