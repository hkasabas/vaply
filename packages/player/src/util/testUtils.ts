import { AnnotationConfig } from "@player/model";
import { DeepPartial } from "@player/util/lang";
import { FilteredByType, typeFind } from "@player/util/unionFilter";
import { merge } from "@player/util/utils";

// ---------- Annotation test utils

/**
 * Uses full annotation `config` and applies partial `changes` to it  and returns new config object.
 *
 * Useful when you need to change just some specific details on an existing config.
 */
export function createAnnotation<T extends AnnotationConfig["type"]>(
  config: FilteredByType<AnnotationConfig, T>,
  changes: DeepPartial<FilteredByType<AnnotationConfig, T>>
) {
  return merge(config, changes);
}

/** Find and return the first annotation of specific `type` from the list.
 * Returns `undefined` if none found.
 */
export function findAnnotation<T extends AnnotationConfig["type"]>(list: AnnotationConfig[], type: T) {
  const annotation = typeFind(list, type);

  if (annotation == null) {
    throw `Annotation with ${type} not found in list`;
  }

  return annotation;
}
