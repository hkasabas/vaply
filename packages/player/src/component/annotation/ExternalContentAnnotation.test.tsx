import { render } from "@testing-library/preact";
import { FunctionComponent, RenderableProps, h } from "preact";

import { Annotation } from "@player/component/annotation/Annotation";
import { commonTests } from "@player/component/annotation/Annotation.test";
import { AnnotationConfig } from "@player/model";
import { findAnnotation } from "@player/util/testUtils";

// ---------- Wrapper component

type WrapperComponentProps = RenderableProps<{ renderAnnotation?: boolean }>;

/**
 * Wrapper component that wraps rendered annotation component and adds external content elements that the annotation can use.
 */
const WrapperComponent: FunctionComponent<WrapperComponentProps> = (props) => {
  return (
    <div>
      {/* set static HTML content that the component can move around */}
      {/* eslint-disable-next-line react/no-danger */}
      <div id="externalContentContainer" dangerouslySetInnerHTML={{ __html: '<div id="externalcontent-annot-1">This is an annotation</div>' }} />

      {(props.renderAnnotation ?? true) && props.children}
    </div>
  );
};
/**
 * Factory functions that renders WrapperComponent with default `renderAnnotation` prop.
 *
 * Wrapper component is rendered outside of our control so we cannot pass new props to it.
 * This way we provide a generic component which in turn renders `WrapperComponent` with desired props.
 */
const wrapperComponentFactory = (renderAnnotation?: boolean) => {
  return (props: WrapperComponentProps) => <WrapperComponent renderAnnotation={renderAnnotation}>{props.children}</WrapperComponent>;
};

describe("external content annotations", () => {
  // --- prepared mock data
  const ANNOTATIONS: AnnotationConfig[] = [
    {
      code: "annot-externalcontent-1",
      type: "externalcontent",
      blocking: true,
      dismissible: true,
      position: { type: "static", static: { right: "5px", top: "50%" } },
      externalcontent: { selector: "#externalcontent-annot-1" },
    },
  ];

  // run common tests

  commonTests(findAnnotation(ANNOTATIONS, "externalcontent"), WrapperComponent);

  // run "externalcontent" specific tests
  describe("externalcontent", () => {
    it("renders", () => {
      const annot = findAnnotation(ANNOTATIONS, "externalcontent");

      // render
      const result1 = render(<Annotation config={annot} />, { wrapper: wrapperComponentFactory() });
      expect(result1.container).toMatchSnapshot();
    });

    it("contains external content", async () => {
      const annot = findAnnotation(ANNOTATIONS, "externalcontent");

      // --- test annotation component mounting
      // render
      const result1 = render(<Annotation config={annot} />, { wrapper: wrapperComponentFactory() });

      // ext. content no longer in it's container
      const prevEl1 = result1.container.querySelector<HTMLDivElement>(`#externalContentContainer #externalcontent-annot-1`);
      expect(prevEl1).toBeNull();

      // ext. content is in annotation
      const newEl1 = result1.container.querySelector<HTMLDivElement>(`.vaply-externalContentAnnotation__container #externalcontent-annot-1`);
      expect(newEl1).not.toBeNull();

      // --- test annotation component UNmounting
      // rerender without annotation
      const result2 = render(<Annotation config={annot} />, { wrapper: wrapperComponentFactory(false) });

      // ext. content no longer in annotation
      const prevEl2 = result2.container.querySelector<HTMLDivElement>(`#externalContentContainer #externalcontent-annot-1`);
      expect(prevEl2).not.toBeNull();

      // ext. content back in it's container
      const newEl2 = result2.container.querySelector<HTMLDivElement>(`.vaply-externalContentAnnotation__container #externalcontent-annot-1`);
      expect(newEl2).toBeNull();
    });
  });
});
