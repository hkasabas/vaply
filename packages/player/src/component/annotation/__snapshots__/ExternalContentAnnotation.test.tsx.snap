// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`external content annotations common applies dimensions 1`] = `
<div>
  <div>
    <div
      id="externalContentContainer"
    />
    <div
      class="vaply-annotation__staticPosition"
      style="top: 50%; right: 5px;"
    >
      <div
        class="vaply-externalContentAnnotation__container"
      >
        <div
          id="externalcontent-annot-1"
        >
          This is an annotation
        </div>
      </div>
      <div
        class="vaply-annotation__closeButton"
      >
        ×
      </div>
    </div>
  </div>
</div>
`;

exports[`external content annotations common applies dimensions 2`] = `
<div>
  <div>
    <div
      id="externalContentContainer"
    >
      <div
        id="externalcontent-annot-1"
      >
        This is an annotation
      </div>
    </div>
    <div
      class="vaply-annotation__staticPosition"
      style="top: 50%; right: 5px; width: 1px; height: 1px;"
    >
      <div
        class="vaply-externalContentAnnotation__container"
      >
        <div
          id="externalcontent-annot-1"
        >
          This is an annotation
        </div>
      </div>
      <div
        class="vaply-annotation__closeButton"
      >
        ×
      </div>
    </div>
  </div>
</div>
`;

exports[`external content annotations common applies dimensions 3`] = `
<div>
  <div>
    <div
      id="externalContentContainer"
    >
      <div
        id="externalcontent-annot-1"
      >
        This is an annotation
      </div>
    </div>
    <div
      class="vaply-annotation__staticPosition"
      style="top: 50%; right: 5px;"
    >
      <div
        class="vaply-externalContentAnnotation__container"
      >
        <div
          id="externalcontent-annot-1"
        >
          This is an annotation
        </div>
      </div>
      <div
        class="vaply-annotation__closeButton"
      >
        ×
      </div>
    </div>
  </div>
</div>
`;

exports[`external content annotations common applies static positioning 1`] = `
<div>
  <div>
    <div
      id="externalContentContainer"
    />
    <div
      class="vaply-annotation__staticPosition"
      style="top: 50%; right: 5px;"
    >
      <div
        class="vaply-externalContentAnnotation__container"
      >
        <div
          id="externalcontent-annot-1"
        >
          This is an annotation
        </div>
      </div>
      <div
        class="vaply-annotation__closeButton"
      >
        ×
      </div>
    </div>
  </div>
</div>
`;

exports[`external content annotations common applies static positioning 2`] = `
<div>
  <div>
    <div
      id="externalContentContainer"
    >
      <div
        id="externalcontent-annot-1"
      >
        This is an annotation
      </div>
    </div>
    <div
      class="vaply-annotation__staticPosition"
      style="top: 1px; right: 1px; bottom: 1px; left: 1px;"
    >
      <div
        class="vaply-externalContentAnnotation__container"
      >
        <div
          id="externalcontent-annot-1"
        >
          This is an annotation
        </div>
      </div>
      <div
        class="vaply-annotation__closeButton"
      >
        ×
      </div>
    </div>
  </div>
</div>
`;

exports[`external content annotations externalcontent renders 1`] = `
<div>
  <div>
    <div
      id="externalContentContainer"
    />
    <div
      class="vaply-annotation__staticPosition"
      style="top: 50%; right: 5px;"
    >
      <div
        class="vaply-externalContentAnnotation__container"
      >
        <div
          id="externalcontent-annot-1"
        >
          This is an annotation
        </div>
      </div>
      <div
        class="vaply-annotation__closeButton"
      >
        ×
      </div>
    </div>
  </div>
</div>
`;
